src/gt4py/analysis/passes.py

```diff
@@ -1194,38 +1194,37 @@ class DemoteLocalTemporariesToVariablesPass(TransformPass):
     """
 
     class CollectDemotableSymbols(gt_ir.IRNodeVisitor):
-        def __init__(self):
-            pass
-
         @classmethod
         def apply(cls, node: gt_ir.StencilImplementation) -> Set[str]:
             collector = cls()
             return collector(node)
 
         def __call__(self, node: gt_ir.StencilImplementation) -> Set[str]:
             assert isinstance(node, gt_ir.StencilImplementation)
-            self.demotables = set(node.temporary_fields)
-            self.seen_symbols: Set[str] = set()
-            self.stage_symbols: Optional[Set[str]] = None
+            self.demotables: Dict[str, str] = {
+                field_name: "" for field_name in node.temporary_fields
+            }
             self.visit(node)
-            return self.demotables
+            return set(self.demotables.keys())
 
         def visit_Stage(self, node: gt_ir.Stage) -> None:
-            self.stage_symbols = set()
-
+            self.stage_name = node.name
             self.generic_visit(node)
 
-            for s in self.stage_symbols:
-                if s in self.seen_symbols:
-                    self.demotables.discard(s)
-                self.seen_symbols.add(s)
-            self.stage_symbols = None
+        def visit_Assign(self, node: gt_ir.Assign) -> None:
+            target_name = node.target.name
+            if target_name in self.demotables:
+                self.demotables[target_name] = self.stage_name
+            self.visit(node.value)
 
         def visit_FieldRef(self, node: gt_ir.FieldRef) -> None:
-            if any(v != 0 for v in node.offset.values()):
-                self.demotables.discard(node.name)
-            assert self.stage_symbols is not None
-            self.stage_symbols.add(node.name)
+            field_name = node.name
+            if field_name in self.demotables:
+                assert len(self.demotables[field_name]) > 0
```

Comment:
Instead of asserting the `len(...) > 0`, the more Pythonic way is probably to make `self.demotables: Dict[str, Optional[str]]` and set it to `None` initially.